File: base/src/main/java/com/windscribe/vpn/serverlist/dao/PingTimeDao.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.vpn.serverlist.dao;

import static androidx.room.OnConflictStrategy.REPLACE;

import androidx.room.Dao;
import androidx.room.Insert;
import androidx.room.Query;

import com.windscribe.vpn.Windscribe;
import com.windscribe.vpn.serverlist.entity.PingTime;

import java.util.List;

import io.reactivex.Completable;
import io.reactivex.Single;

@Dao
abstract public class PingTimeDao {


    public Completable addPingTime(PingTime pingTime) {
        return addPing(pingTime);
    }

    @Query("Select * from PingTime")
    public abstract Single<List<PingTime>> getAllPings();

    @Query("Select ping_id from PingTime where ping_time =:pingTime and isPro=:pro")
    public abstract Single<Integer> getFreePingIdFromTime(boolean pro, int pingTime);

    @Query("Select MIN(ping_time) from PingTime where ping_time!= -1 and static=0 limit 1")
    public abstract Single<Integer> getLowestPing();

    @Query("Select MIN(ping_time) from PingTime where ping_time!= -1 and isPro=:pro and static=0 limit 1")
    public abstract Single<Integer> getLowestPingForFreeUser(boolean pro);

    public Single<Integer> getLowestPingId() {
        boolean freeUser = Windscribe.getAppContext().getPreference().getUserStatus() == 0;
        if (freeUser) {
            return getLowestPingForFreeUser(false).flatMap(time -> getFreePingIdFromTime(false, time));
        } else {
            return getLowestPing().flatMap(this::getPingIdFromTime);
        }
    }

    @Query("Select ping_id from PingTime where ping_time =:pingTime")
    public abstract Single<Integer> getPingIdFromTime(int pingTime);

    @Insert(onConflict = REPLACE)
    abstract Completable addPing(PingTime pingTime);
}

File: base/src/main/java/com/windscribe/vpn/serverlist/dao/PingTimeDao.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.vpn.serverlist.dao

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy.Companion.REPLACE
import androidx.room.Query
import com.windscribe.vpn.Windscribe
import com.windscribe.vpn.serverlist.entity.PingTime
import kotlinx.coroutines.flow.Flow

@Dao
abstract class PingTimeDao {

    @Query("Select * from PingTime")
    abstract suspend fun getAllPingsAsync(): List<PingTime>

    @get:Query("Select * from PingTime")
    abstract val allPingsAsStateFlow: Flow<List<PingTime>>

    suspend fun getLowestPingIdAsync(): Int {
        val freeUser = Windscribe.appContext.preference.userStatus == 0
        return if (freeUser) {
            val time = getLowestPingForFreeUserAsync(false)
            getFreePingIdFromTimeAsync(false, time)
        } else {
            val time = getLowestPingAsync()
            getPingIdFromTimeAsync(time)
        }
    }

    @Query("Select MIN(ping_time) from PingTime where ping_time!= -1 and static=0 limit 1")
    abstract suspend fun getLowestPingAsync(): Int

    @Query("Select MIN(ping_time) from PingTime where ping_time!= -1 and isPro=:pro and static=0 limit 1")
    abstract suspend fun getLowestPingForFreeUserAsync(pro: Boolean): Int

    @Query("Select ping_id from PingTime where ping_time =:pingTime and isPro=:pro")
    abstract suspend fun getFreePingIdFromTimeAsync(pro: Boolean, pingTime: Int): Int

    @Query("Select ping_id from PingTime where ping_time =:pingTime")
    abstract suspend fun getPingIdFromTimeAsync(pingTime: Int): Int

    @Insert(onConflict = REPLACE)
    abstract suspend fun addPing(pingTime: PingTime)
}